### PR TITLE
Extend menu commands 

### PR DESCRIPTION
**Pull request by john-traas**
Updates the menu-command factory to make it possible to toggle menu commands. 

Fixes https://github.com/Lundalogik/crm-feature/issues/4130

### CODE REVIEW

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:46]: Ok, so:

- if `node && node.type.name === nodeType.name`, but `nodeType.name` is not `'heading'`, return `true`
- if `node && node.type.name === nodeType.name`, and `nodeType.name` _is_ `'heading'`, but `level` is falsy, return `false`
- if `node && node.type.name === nodeType.name`, and `nodeType.name` _is_ `'heading'`, and `level` is truthy, return `true` if `node.attrs.level` and `level` are the same

Correct? 😅

Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:46]: Yes. 😅 

This set of conditions should never happen as a `heading` must have a `level`:
if `node && node.type.name === nodeType.name`, and `nodeType.name` is `'heading'`, but `level` is falsy, return false

Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:46]: We determine the `node.type` of the `current node` to show it is `active`. If the `node.type` is a `'heading'`, the `level` must be the same as the one on the menu item that was clicked so the relevant menu item shows as selected (active).

File: src/components/text-editor/prosemirror-adapter/menu/types.ts
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/menu/types.ts:19]: I'm getting errors like these when using underline and strikethrough:

image

Underline and strikethrough looks like they are working in the editor, but as soon as you have any part of the text formatted with either of them, the output stops updating.

I don't think we should spend time debugging this until we know whether underline and strikethrough are in the set of MVP features though.

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts
Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:46]: Ah, yeah, "setActiveMethodForNode". Ok, that makes sense then 👍 🙂

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:188]: I'm not sure if markdown supports nested lists. The current behaviour seems to be fairly similar to how the list buttons work in the editor here on GitHub. In the future, I imagine we might want more advanced behaviour, at least when the output mode is html, but if so, that's for another time.

👍

File: src/components/text-editor/prosemirror-adapter/menu/types.ts
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/menu/types.ts:21]: This PR is broken into commits in quite a weird way, imo. I'd rather see that any changes that were already needed before adding new types here were made first, in one or more commits. (I'm guessing one commit, since it's basically fixing one problem: "be able to toggle stuff on and off".) Then any new features, like support for underline and strikethrough, should be added in their own commits.

But, unless underline and strikethrough are considered MVP, we shouldn't add them now.

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:155]: I agree that it is a bit annoying, and it's the same for the lists in the current state. I have some thoughts about how to implement a better experience but felt based on the bet for the cycle, that it wasn't a priority at the moment.

File: src/components/text-editor/prosemirror-adapter/menu/types.ts
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/types.ts:19]: It's a very small fix for these two. The marks for underline and strikethrough aren't in the default markdown serialiser that prosemirror provides.

Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/menu/types.ts:19]: OK, if you know what's needed, and it doesn't turn into a rabbit hole (aka, you can fix it before the meeting at 16:00), go ahead 😊

File: src/components/text-editor/prosemirror-adapter/menu/types.ts
Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/types.ts:21]: I can re-arrange the commits in an order that makes more sense.

File: src/components/text-editor/prosemirror-adapter/menu/types.ts
Comment (adrianschmidt) [reply src/components/text-editor/prosemirror-adapter/menu/types.ts:19]: (To clarify, I don't mean to take the role of some ultimate dictator, deciding what you guys can and cannot do. My intention is just to provide a firm guiding hand 😊)

File: src/components/text-editor/utils/markdown-converter.ts
Comment (adrianschmidt) [inline src/components/text-editor/utils/markdown-converter.ts:25]: Have we decided anything on the markdown vs html output question?

I'm asking since we seem to be mixing html (`<u>undeline</u>`) and markdown (`~~strikethrough~~`) here…? I know vanilla markdown doesn't actually support underline, while some flavours support interpreting enclosing single underscores as underline (which I've always thought makes sense 🤷‍♂️).

Comment (adrianschmidt) [reply src/components/text-editor/utils/markdown-converter.ts:25]: Oh… `customMarkdownSerializer`… so I guess this is when outputting markdown then?

Comment (john-traas) [reply src/components/text-editor/utils/markdown-converter.ts:25]: This is when `markdown` is set as the `contentType`, but your question does have me thinking further about this.

Comment (adrianschmidt) [reply src/components/text-editor/utils/markdown-converter.ts:25]: Are we outputting markdown or html by default?

I would assume the main focus right now is html anyway, since that's presumably what the email feature needs?

Comment (john-traas) [reply src/components/text-editor/utils/markdown-converter.ts:25]: We are outputting markdown by default but that can be changed by modifying one line in the `prosemirror-adapter`

Comment (adrianschmidt) [reply src/components/text-editor/utils/markdown-converter.ts:25]: Should we hold off on including the underline button?

I mean, it will work if the output is formatted with `limel-markdown`, as that can handle mixed markdown and html. But we're not actually outputting pure markdown when there's something that's been underlined. So I'm thinking we might not want to include the underline button in the default set of menu buttons when the output mode is markdown…

But we don't want to spend any time on including it when in html mode and excluding it in markdown mode. Not right now.

Comment (adrianschmidt) [reply src/components/text-editor/utils/markdown-converter.ts:25]: I'm thinking I can just move that commit to a separate branch, and you can open a PR for that commit only, and then we can leave that to rest until we're ready to look at it.

Comment (john-traas) [reply src/components/text-editor/utils/markdown-converter.ts:25]: I think that sounds like a good solution.

File: src/components/text-editor/prosemirror-adapter/menu/menu-command.spec.tsx
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/menu/menu-command.spec.tsx:1]: I've tested it a little bit in Safari on an iPhone via BrowserStack, and that seems to work well enough. There seemed to be some odd behaviour around adding new lines inside a code block, but honestly, I can't see people adding a whole lot of code blocks via their phones…

**Review by adrianschmidt — APPROVED** [review]

File: src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts
Comment (adrianschmidt) [inline src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:139]: Oh, yeah, I didn't want to sidetrack the current cycle into a massive refactoring effort, but once we have achieved MVP, I think we will need to go over the whole text-editor component and refactor a bunch of parts, just to break stuff out into named functions, because following stuff like this is not easy, and a few months down the line, when there's no one who has it fresh in their memory that you can ask, it's going to be a nightmare.

(Not just a comment about this code specifically, there seems to be a lot of deeply nestled code since earlier too.)

Comment (john-traas) [reply src/components/text-editor/prosemirror-adapter/menu/menu-commands.ts:139]: I definitely agree!

**Pull request merged**
